scene/3d/collision_shape_3d.cpp

```diff
@@ -220,6 +227,37 @@ void CollisionShape3D::_update_debug_shape() {
 	mi->set_mesh(mesh);
 	add_child(mi);
 	debug_shape = mi;
+
+	// Visible Faces
+	Ref<ArrayMesh> mesh_faces;
```

Comment:
The logic for drawing shapes has been moved to `CollisionObject3D` and is now doing calls to the `RenderingServer` directly instead of adding mesh instances in the tree.

So after a rebase, `debug_shape_faces` will have to move to `CollisionObject3D`, see:
https://github.com/godotengine/godot/blob/ade4e9320a6ca403b8053fe5828d3f9ce809338c/scene/3d/collision_object_3d.cpp#L340

And also:
-Same as for the editor gizmo, it could use some simpler logic by making `get_debug_arraymesh_faces()` virtual just like `get_debug_mesh()` does.
-Options for faces enabled and custom colors can be check from project settings
`GLOBAL_GET("debug/shapes/collision/draw_3d_faces")`
`GLOBAL_GET("debug/shapes/collision/use_random_color_for_3d_faces")`
-It needs one more option in project settings to set the default color, defined in scene tree as well:
GLOBAL_DEF("debug/shapes/collision/shape_color_3d_faces", Color(0.0, 0.6, 0.7, 0.42));

Reply:
Also, I noticed a material is created for every instance. This will be a bottleneck in large scenes.

Instead, vertex colors should be used to tint each shape by giving it vertex colors when generating the mesh, and displaying those as albedo (similar to what is done for Light3D gizmo lines and billboard icons).